### PR TITLE
Decrease memory overhead for contiguous query row ranges.

### PR DESCRIPTION
**Pull request by aoblebea**
Clamped query row ranges to array size (when possible) and created interval_expander class to store row intervals while largely mirroring the api of the previous row vector. Should use vastly less memory when users query large contiguous row regions (even if they are not specified as intervals). Will have something of a memory and performance overhead on large queries with very few contiguous regions.

### CODE REVIEW

File: src/main/cpp/src/config/variant_query_config.cc [hunk 1]
```diff
@@ -258,15 +186,138 @@ void VariantQueryConfig::validate(const int rank) {
...
+
+  for(auto& e : vec) {
+    int64_t first = e.first, second = e.second;
+    // make sure interval isn't backwards
+    if(e.first <= e.second) {
+      first = e.first;
+      second = e.second;
+    }
+    else {
+      first = e.second;
```
Comment (nalinigans) [inline src/main/cpp/src/config/variant_query_config.cc:207]: Just a stylistic suggestion - can you have `} else {` in the same line?
```
if (this) {
  do_this();
} else {
  do that(); 
}
```

File: src/main/cpp/src/config/variant_query_config.cc [hunk 2]
```diff
@@ -258,15 +186,138 @@ void VariantQueryConfig::validate(const int rank) {
...
+      second = e.first;
+    }
+
+    // make sure interval is in bounds (on the left)
+    if(first < 0) {
```
Comment (nalinigans) [inline src/main/cpp/src/config/variant_query_config.cc:212]: If `vec` can be forced into type `std::pair<uint64_t, uint64_t>`, you can avoid this check with first.

File: src/main/cpp/src/config/variant_query_config.cc [hunk 3]
```diff
@@ -258,15 +186,138 @@ void VariantQueryConfig::validate(const int rank) {
...
+      first = 0;
+    }
+    if(second >= 0) {
```
Comment (nalinigans) [inline src/main/cpp/src/config/variant_query_config.cc:215]: Same as comment in line 212.

**Review by nalinigans — CHANGES_REQUESTED** [review]
Very nice @aoblebea. Just minor comments to address.

File: src/main/cpp/src/config/variant_query_config.cc (hunk 2 above)
Comment (aoblebea) [reply src/main/cpp/src/config/variant_query_config.cc]: Kept vec as int64_t to be consistent with TileDBRowRange, but removed check to allow negative values.

File: src/main/cpp/include/config/variant_query_config.h [hunk 4]
```diff
@@ -55,6 +55,60 @@ class UnknownQueryAttributeException : public std::exception {
...
+    // sets total_size_before field of each interval in intervals
+    void index();
+    // check if array row is queried
+    bool is_queried_row(int64_t row) const;
+    // NOTE: triggers an assertion if row is out of bounds
+    int64_t get_array_row_from_query_row(int64_t row) const;
+    // equivalent to get_array_row_from_query_row
+    int64_t operator[](int64_t idx) const;
+    // NOTE: triggers an assertion if row is out of bounds
+    int64_t get_query_row_from_array_row(int64_t row) const;
```
[the remark this answers is not in the document]
Comment (aoblebea) [reply src/main/cpp/include/config/variant_query_config.h:104]: get_array_row... vs get_query_row...

File: src/main/cpp/include/config/variant_query_config.h [hunk 5]
```diff
@@ -55,6 +55,60 @@ class UnknownQueryAttributeException : public std::exception {
...
+      // number of rows in interval
+      inline int64_t size() const { return second - first + 1; }
+      inline int64_t span() const { return second - first; }
+      // total number of rows in interval and all preceeding
+      inline int64_t size_inclusive() const { return second - first + 1 + total_size_before; }
+      bool intersects(const interval& r) const {
+        return second >= r.first && first <= r.second;
+      }
+      // bounding interval/generalized intersection
+      interval operator+(const interval& r) const {
```
[the remark this answers is not in the document]
Comment (aoblebea) [reply src/main/cpp/include/config/variant_query_config.h:76]: operator+ is implementing generalized intersection, but it should only be called on overlapping intervals by update_rows

**Review by nalinigans — COMMENTED** [review]
Looks good @aoblebea . Did you run any benchmark tests querying by rows and by ranges against the 1000g samples workspace with and without these changes?

**Comment by nalinigans** [comment]
@aoblebea I would like to bring this PR into the release we will be making soon. Two questions :

1. Did you run any benchmark tests querying by rows and by ranges against the 1000g samples workspace with and without these changes?
2. Would it be possible to write cunit to cover and target query row ranges specifically? You could maybe write junit tests too - see https://github.com/GenomicsDB/GenomicsDB/blob/master/src/test/java/org/genomicsdb/reader/GenomicsDBQueryTest.java

**Comment by aoblebea** [comment]
> @aoblebea I would like to bring this PR into the release we will be making soon. Some questions :
> 
> 1. Did you run any benchmark tests querying by rows and by ranges against the 1000g samples workspace with and without these changes?
> 2. Would it be possible to write cunit to cover and target query row ranges specifically? You could maybe write junit tests too - see https://github.com/GenomicsDB/GenomicsDB/blob/master/src/test/java/org/genomicsdb/reader/GenomicsDBQueryTest.java
> 3. Does this fix Issue [GenomicsDB api returns a null list when querying for variant calls with no rows specified #157](https://github.com/GenomicsDB/GenomicsDB/issues/157)?

1. Just some failure cases of the old solution with huge row ranges.
2. I can do that
3. I don't remember, I'll have to check

File: src/main/cpp/include/config/variant_query_config.h (hunk 5 above)
Comment (nalinigans) [inline src/main/cpp/include/config/variant_query_config.h:76]: Can you write a simple test with intersecting ranges to exercise this code and quiet codecov please?

**Review by nalinigans — CHANGES_REQUESTED** [review]

**Review by mlathara — APPROVED** [review]
looks good

**Comment by nalinigans** [comment]
Changes look good. Performance with this row range implementation has improved compared to before. Thanks @aoblebea

**Review by nalinigans — APPROVED** [review]
looks good.

**Pull request merged**